pkg/driver/node.go

```diff
@@ -283,8 +285,7 @@ func (d *nodeService) NodeUnstageVolume(ctx context.Context, req *csi.NodeUnstag
 	// returns the device name, reference count, and error code
 	dev, refCount, err := d.mounter.GetDeviceNameFromMount(target)
 	if err != nil {
-		msg := fmt.Sprintf("failed to check if volume is mounted: %v", err)
-		return nil, status.Error(codes.Internal, msg)
```

Comment:
per the CSI spec: _If the volume corresponding to the volume_id is not staged to the staging_target_path, the Plugin MUST reply 0 OK._  Returning an error here is a bug, instead, we should log an error message.

If `GetDeviceNameFromMount` returns an error, it will always return `refCount = 0` which triggers https://github.com/torredil/aws-ebs-csi-driver/blob/205c734ce383c9ed6418279fa0a71377f6b8e751/pkg/driver/node.go#L296. Thus, we do not attempt to ever unstage/unmount volumes that are not staged.

Reply:
`GetDeviceNameFromMount` returning an error ALWAYS indicates the volume is always not staged? Shouldn't we differentiate between other errors?

Reply:
> returning an error ALWAYS indicates the volume is not staged?

Yes, if there is no volume ID found at the target path then the volume is not staged and `refCount = 0`.

> Shouldn't we differentiate between other errors (or just directly check if it is staged)?

We directly check if the volume is staged at a given path.

[d.mounter.GetDeviceNameFromMount(target)](https://github.com/torredil/aws-ebs-csi-driver/blob/205c734ce383c9ed6418279fa0a71377f6b8e751/pkg/driver/node.go#L286) -> [proxyMounter.GetDeviceNameFromMount(mountPath)](https://github.com/torredil/aws-ebs-csi-driver/blob/0c117785e5e4cbdbc593ecea88d438672fe0a3c8/pkg/driver/mount_windows.go#L51) which [checks if the volume is staged](https://github.com/torredil/aws-ebs-csi-driver/blob/205c734ce383c9ed6418279fa0a71377f6b8e751/pkg/mounter/safe_mounter_windows.go#L222) by retrieving the volume ID at the given target path.

Reply:
> Yes, if there is no volume ID found at the target path then the volume is not staged and `refCount = 0`.

Can you demonstrate why this is the ONLY POSSIBLE error return from `GetDeviceNameFromMount`. I can see another possible error right there in the function even: https://github.com/torredil/aws-ebs-csi-driver/blob/0c117785e5e4cbdbc593ecea88d438672fe0a3c8/pkg/driver/mount_windows.go#LL47-L50C3 - this is the part I don't understand.

Reply:
Newest revision addresses this feedback by returning a nil error in `GetDeviceNameFromMount` if the target path does not exist.